### PR TITLE
Add expo client:ios command for ad hoc builds

### PR DESCRIPTION
**Pull request by fson**
# Why

This workflow will allow creating custom builds of the Expo Client and installing them on an iOS device.

# How

Added the command `client:ios` which shares a lot of code with the `build:ios` command.

The command:
* authenticates the user with the Apple Developer Portal using Apple ID and selects the team ID to use.
* authenticates with Expo (**optional**)
* generates a stable bundle identifier for the custom Expo Client from the Apple team ID and adds it to the Apple Dev Portal. Bundle ID is created with the following formula: `bundleIdentifier = 'dev.expo.client.' + base32(sha224(teamId))`. E.g.: `dev.expo.client.baepmtta2wext7fwo3ew5sjye4g6uqsellxpzu5e434nnw`
* allows the user to choose which distribution certificate to use: Screenshot 2019-03-15 at 17 39 35

* allows the user to choose which push key to use
* asks for an email address to use for build notifications (if not logged in with Expo)
* lists existing devices on Apple Dev Portal and asks if the user wants to add new devices
* posts the distribution certificate and list of devices to the Expo API to initiate the Turtle build
* if "add new devices" selected, shows the URL for adding the device

# Test Plan

Ran `expo client:ios`

Screenshot 2019-03-15 at 17 22 03Screenshot 2019-03-15 at 17 22 19

### CODE REVIEW

File: packages/expo-cli/package.json [hunk 1]
```diff
@@ -67,6 +65,7 @@
     "@expo/spawn-async": "1.3.0",
     "axios": "0.18.0",
     "babel-runtime": "6.26.0",
+    "base32.js": "^0.1.0",
```
Comment (dsokal) [inline packages/expo-cli/package.json]:
```suggestion
    "base32.js": "0.1.0",
```

File: packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js [hunk 2]
```diff
@@ -40,7 +40,8 @@ function _getAppleIdFromParams({ appleId }) {
   }
 }
 
-async function _promptForAppleId({ appleId }) {
+async function _promptForAppleId(options) {
+  let appleId = options.appleId;
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js]: Why did you move it here?

File: packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js [hunk 3]
```diff
@@ -52,22 +53,32 @@ async function _promptForAppleId({ appleId }) {
...
 
-  const appleIdQuestions = [
-    {
-      type: 'input',
-      name: 'appleId',
-      message: `Apple ID:`,
-      validate: nonEmptyInput,
-      when: () => !appleId,
-    },
+  if (!appleId) {
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js]: Why? Note there is `when: () => !appleId,` in the question object, so this `if` is redundant.

File: packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js [hunk 4]
```diff
@@ -52,22 +53,32 @@ async function _promptForAppleId({ appleId }) {
...
       validate: nonEmptyInput,
     },
-  ];
-  return { appleId, ...(await prompt(appleIdQuestions)) };
+    {
+      nonInteractiveHelp:
+        'Pass your Apple ID password using the EXPO_APPLE_PASSWORD environment variable',
+    }
+  );
+  return { appleId, appleIdPassword };
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js]: Changes in this function seem redundant to me. You could only add `nonInteractiveHelp` fields to the question objects.

File: packages/expo-cli/src/commands/build/ios/appleApi/distributionCert.js [hunk 5]
```diff
@@ -26,7 +26,9 @@ const createManager = ({ appleId, appleIdPassword, team }) => ({
     } catch (err) {
       const resultString = get(err, 'rawDump.resultString');
       if (resultString && resultString.match(/Maximum number of certificates generated/)) {
-        log.error(APPLE_DIST_CERTS_TOO_MANY_GENERATED_ERROR);
+        let error = new Error(APPLE_DIST_CERTS_TOO_MANY_GENERATED_ERROR);
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build/ios/appleApi/distributionCert.js]: I prefer to use `const` for things that don't change.

File: packages/expo-cli/src/commands/build/ios/credentials/revoke.js [hunk 6]
```diff
@@ -30,6 +30,7 @@ async function _askWhatToRevoke(choices) {
     type: 'checkbox',
     name: 'ids',
     message: `What would you like to revoke?`,
+    pageSize: Infinity,
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build/ios/credentials/revoke.js]: 👍

File: packages/expo-cli/src/commands/client/index.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+import prompt from '../../prompt';
+
+export default program => {
+  program
+    .command('client:ios')
+    .option(
+      '--apple-id <login>',
+      'Apple ID username (please also set the Apple ID password as EXPO_APPLE_PASSWORD environment variable).'
+    )
+    .asyncAction(async options => {
```
Comment (dsokal) [inline packages/expo-cli/src/commands/client/index.js]: I prefer to keep command files clean - so that they only contain the definition of a command and the "business logic" is kept in a separate file. Consider this as a suggestion, I don't have a strong opinion on this.

File: packages/expo-cli/src/commands/client/index.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      let pushKey = await selectPushKey(context);
+
+      let email;
+      if (user) {
+        email = user.email;
+      } else {
+        ({ email } = await prompt({
+          name: 'email',
+          message: 'Please enter an email address to notify, when the build is completed:',
+          default: user ? user.email : undefined,
```
Comment (dsokal) [inline packages/expo-cli/src/commands/client/index.js]: I guess it will be always `undefined`.

File: packages/expo-cli/src/commands/client/index.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+          filter: value => value.trim(),
+          validate: value => (/@/.test(value) ? true : "That doesn't look like a valid email."),
```
Comment (dsokal) [inline packages/expo-cli/src/commands/client/index.js]: I think we need a better validation here.

File: packages/expo-cli/src/commands/client/index.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            }`
+          )
+        );
+        log.newLine();
+      } else {
+        log(
+          chalk.green(`Build started! You will receive an email to ${email} when it's finished.`)
+        );
+      }
+    });
```
Comment (dsokal) [inline packages/expo-cli/src/commands/client/index.js]: This function seems to be too complicated and does things at different levels of abstraction.

File: packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js (hunk 4 above)
Comment (fson) [reply packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js]: `nonInteractiveHelp` is not a property of the Inquirer.js questions, but a separate option handled by the `prompt` function. That's why I split the prompt to two separate calls.

File: packages/expo-cli/src/commands/client/index.js (hunk 8 above)
Comment (fson) [reply packages/expo-cli/src/commands/client/index.js]: Uh, thanks. I think I changed the logic at some point and forgot to remove this line.

File: packages/expo-cli/src/commands/client/index.js (hunk 9 above)
Comment (fson) [reply packages/expo-cli/src/commands/client/index.js]: The only way to reliably validate an email address is to try sending email to it. Failing that, it's better to keep the validation as simple as possible to avoid false negatives. (E.g. some web forms don't allow addresses like `myemail+label@example.com` or the more obscure TLDs.) Even the full [regular expression to validate an email address according to RFC822](http://www.ex-parrot.com/~pdw/Mail-RFC822-Address.html) won't help if I have a typo in any individual component of the address.

Comment (dsokal) [reply packages/expo-cli/src/commands/client/index.js]: Well, yes, but `@` is a valid email according to this condition. We could at least check whether there is anything before and after the `@` character.

File: packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js (hunk 2 above)
Comment (fson) [reply packages/expo-cli/src/commands/build/ios/appleApi/authenticate.js]: I don't remember, I guess I needed the `options` object in some iteration of this. I've reverted this change.

File: packages/expo-cli/src/commands/client/index.js (hunk 10 above)
Comment (fson) [reply packages/expo-cli/src/commands/client/index.js]: I have now moved the API call to a separate function. The rest I prefer to keep in the same function: it's easy to see the whole flow of the command at a single glance without having to look in many files.

File: packages/expo-cli/src/commands/build/ios/appleApi/distributionCert.js (hunk 5 above)
Comment (fson) [reply packages/expo-cli/src/commands/build/ios/appleApi/distributionCert.js]: Ok!

File: packages/expo-cli/src/commands/client/index.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        context,
+        distributionCert,
+        udids,
+        addUdid,
+        email,
+      });
+
+      log.newLine();
+      if (addUdid) {
+        urlOpts.printQRCode(result.registrationUrl);
```
[the remark this answers is not in the document]
Comment (fson) [reply packages/expo-cli/src/commands/client/index.js]: Nice work! The QR codes will look much better now.

File: packages/expo-cli/src/commands/client/index.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+          chalk.green(
+            `Please visit this link on your device to register its UDID and download Expo Client:\n. ${
+              result.registrationUrl
+            }`
+          )
+        );
+        log.newLine();
+      } else {
+        log(
+          chalk.green(`Build started! You will receive an email to ${email} when it's finished.`)
```
[the remark this answers is not in the document]
Comment (fson) [reply packages/expo-cli/src/commands/client/index.js]: 👍
